### PR TITLE
[DPE-5637, DPE-5276] Implement expose-external config option with values false (clusterip), nodeport and loadbalancer

### PR DESCRIPTION
**Pull request by shayancanonical**
## Prerequisite
Need to merge https://github.com/canonical/data-platform-workflows/pull/244 for metallb support in CI

## Issue
We have outlined the approach to expose our K8s charms externally in DA122

Summary:
- we will avoid using `external-node-connectivity` provided in data_interfaces in K8s charms
- instead, we will provide a new config option `expose-external` with values `false` (for clusterip svc), `nodeport` (for nodeport svc) and `loadbalancer` (for loadbalancer svc)
- we will asynchronously update the endpoints in relation databags (data_interfaces) when we are able to connect to router with the created svc using a socket

## Solution
Implement the contents of the spec

## Testing
- tested locally using metallb
- testing on EKS

### CODE REVIEW

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py:178]: filed a juju bug: https://bugs.launchpad.net/juju/+bug/2084711, havent heard back yet

essentially, we have included deletion + recreation of service as a workaround until we get help from juju to determine what may be happening

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py:270]: prior to this PR, we were only providing the `host:node_port` of one unit in the application. upon discussions, we realized that we would need to provide all `host:node_port`s where the units are scheduled. we are unable to determine nodes where units are deployed without the peer relation which provides all the available/active units

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py:65]: we want to ensure that endpoints are actually connectable before we return them to the client (at the end of the event, we update data_interfaces and trigger a relation_changed on the client). this is similar to ensuring the mysqlrouter is running before event handler exits when enabling mysqlrouter

on cloud certain clouds, it can take a while for the service to be correctly created (along with the loadbalancer). will change so `_wait_until_service_reconciled` is only called on config-changed event so we are not potentially checking for every event

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py:196]: please see [the response here](https://github.com/canonical/mysql-router-k8s-operator/pull/328#discussion_r1805107153). happy to answer followups/discuss further

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py:288]: as commented in another message, i will add a check that `_wait_until_service_reconciled` is only called when the event is `config-changed`. this `_wait_until_service_reconciled` ensures a non-empty host/port that is connectable

 if the host_port is empty during other events, we will update the relation databag with an empty host/port (as there is none available at the moment)

did you mean to discuss a specific use case?

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py:329]: im vm, we are evaluating `external_connectivity` in `database_provides.py`, whereas in k8s we are providing service host:ports (without the need for external_connectivity). the paradigms are a bit different, and so i am finding it difficult to figure out how to unify them

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py]: addressed in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3

File: src/abstract_charm.py
Comment (shayancanonical) [reply src/abstract_charm.py]: only calling this on ConfigChangedEvent in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3

File: src/abstract_charm.py
Comment (shayancanonical) [reply src/abstract_charm.py:345]: moved update endpoints near the reconcile_service call in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py]: addressed in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py]: addressed in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py]: addressed in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py]: addressed in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3. no longer necessary if we `wait_for_service_to_be_reconciled()` only upon ConfigChangedEvent

File: tests/integration/juju_.py
Comment (shayancanonical) [reply tests/integration/juju_.py]: addressed in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3

File: pyproject.toml
Comment (shayancanonical) [reply pyproject.toml]: addressed in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py]: great catch! updated SANs in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3

File: tests/unit/conftest.py
Comment (shayancanonical) [reply tests/unit/conftest.py]: addressed in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3 by avoiding mocking the client

File: tests/unit/conftest.py
Comment (shayancanonical) [reply tests/unit/conftest.py]: avoided in ffc728a96621f8d8529a7ae3f7f562dcae2bdca3

File: .github/workflows/ci.yaml
Comment (shayancanonical) [reply .github/workflows/ci.yaml]: yes, will open a PR in dpw shortly

File: .github/workflows/ci.yaml
Comment (taurus-forever) [inline .github/workflows/ci.yaml:99]: Please revert before merging.

File: src/kubernetes_charm.py
Comment (taurus-forever) [reply src/kubernetes_charm.py:65]: Sleeping up to 5 minutes might produce more issues.
Do we have other options here? Time for Pebble notices?

Comment (delgod) [reply src/kubernetes_charm.py:65]: can we use defer here?

Comment (shayancanonical) [reply src/kubernetes_charm.py:65]: i am leaning towards carl's second option of checking if the service is created before updating endpoints
however, this means that in the worst case, we may wait `update-status-hook-interval` until we update the endpoints in the databag (even though the endpoint is ready much earlier)

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py:193]: set app status to blockedstatus in ab4977647b51d285006bb26c8638090fb7926311if the value is invalid

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py:65]: refactored in ab4977647b51d285006bb26c8638090fb7926311. now, we only update_endpoints if they are connectable. otherwise we will do so in a future event when they become connectable

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py]: no, i did not yet test with EKS or GKE. i dont believe that testing with these platforms should necessarily be a blocker for this PR

we tested on AKS, and this issue did not manifest itself in AKS

Comment (shayancanonical) [reply src/kubernetes_charm.py]: i agree, but we are unable to run integration tests without deleting and recreating (as tests experience flicker of the juju client and fail with a `Bad file descriptor` error). @paulomach please share your thoughts when you are able

Comment (paulomach) [reply src/kubernetes_charm.py]: As @shayancanonical , I validated the flaky behavior independently of a charm, but saw no issue in AKS.
Independently, this does not seems to be an issue in the charm/lightkube.
So let's not block the PR, since this can be refactored once we have better understanding or a fix.

File: src/kubernetes_charm.py
Comment (paulomach) [reply src/kubernetes_charm.py:65]: On the server we are erroring out the hook after 60 seconds (generally ok for a `ClusterIP` service) with a timeout error.
So hook retry will eventually resolve it (or a `juju resolve`), if it's just a timing issue.

So, if endpoints are not connectable we will keep erroring out indefinitely. IMO it's the correct approach, even though ugly, since it renders the server useless if a client cannot use it, while allowing the manual retries (with `juju resolve`)

File: src/kubernetes_charm.py
Comment (paulomach) [inline src/kubernetes_charm.py:191]: same as above

File: src/kubernetes_charm.py
Comment (paulomach) [inline src/kubernetes_charm.py:189]: Any reason to not merge `_apply_service` into `_reconcile_service`?

**Review by paulomach — APPROVED** [review]
I like the tests. If feasible it would be nice to have proper config validation, but that can be done later, given the timing we want to achieve.
There are some other non-blocking comments

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py:65]: i actually did refactor the code so that it no longer blocks for 5mins. instead, when we make a change to the service, we will update the endpoints in the relation databag when the next hook that runs (or this hook) where the service is connectable. i will remove the unused variable shortly

File: config.yaml
Comment (shayancanonical) [reply config.yaml:8]: @paulomach any ideas for alternatives? `expose-external: none`? `expose-external: clusterip`?

Comment (shayancanonical) [reply config.yaml:8]: additionally, the possible values are `false` and `nodeport` in [kafka-k8s](https://github.com/canonical/kafka-k8s-operator/blob/main/config.yaml#L112). it may not be a good idea to introduce inconsistencies

Comment (paulomach) [reply config.yaml:8]: oh bummer, ok maybe for another time.

File: src/kubernetes_charm.py
Comment (paulomach) [reply src/kubernetes_charm.py:65]: on refresh, does juju retry the hook after, before or lost the errored event?

I'm ok with retrying in next hook

File: src/kubernetes_charm.py
Comment (paulomach) [reply src/kubernetes_charm.py:101]: yes, hence not blocking it. Once we reach (if ever) more configs (>3) we should do it

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py:65]: actually, what error are you guys referring to?

if related to service connectivity check, the check itself does not raise any errors. instead, it checks if the service socket is connectable, and if so, sets endpoints in the databag. else, no-ops and will check during the next event that calls `reconcile`

File: config.yaml
Comment (paulomach) [reply config.yaml:8]: Interestingly, it seems it was a Marc nitpick also that Mykola did not responded to in the original [kafka spec](https://docs.google.com/document/d/1JNrThrEULwA2nxi5ZQ9-mkbEGi5SS6N7Ku7MIyfkWbo/edit?disco=AAABUX0Z-sU)

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py]: no reason, merged in [f938351](https://github.com/canonical/mysql-router-k8s-operator/pull/328/commits/f9383516983d97e4d0cd34eb0c020ead40b08539)

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py]: @paulomach @carlcsaposs-canonical i have some additional info on this topic

`lightkube_client.apply` (or any other lightkube client method) is a blocking method (does not return until the change has been made). as far as i can see, we cannot make this call non-blocking. this means that config-changed will run for a while until the service has been created and returns to the lightkube client

however, i also tested deleting a pod while the lightkube client was waiting for the service to be created. with the code as it is right now, the charm recovered. the service continued to be created once the lightkube client disconnected and pod restarted. upon restart, the endpoints get updated in the databag correctly when the next hook runs. i think that it is not worth overthinking here for refresh cases, as we can recover from them correctly

Comment (shayancanonical) [reply src/kubernetes_charm.py]: you're right, the api request blocks until the resource has been created

im trying to explain why there is no issue at all - the api request blocks and with the current implementation, there is no downside to the pod getting refreshed/deleted/rescheduled. the reconcile approach ensures that the databag is correctly updated with the new endpoints during the next event

i think it would be irrelevant if we used apply (with server-side apply) or create, since we would only be applying/creating resources from the leader unit of the router app

**Comment by theoctober19th** [comment]
Great work @shayancanonical!

I'm about to do a similar feature in `kyuubi-k8s` and going through this PR I had a few queries. Looking at the spec, it is my understanding that whenever the service needs to be created / deleted ("reconciled", in other words), the config-changed hook would simply return back after `kubectl.apply` and not wait for the service to be completely created.

When do we check whether the "reconciliation" was successful? In the PR I saw this has been checked on the `config-changed` hook itself, but does that mean the endpoints would not be updated until the config is changed sometime again in the future? Shouldn't we check for whether reconciliation was successful or not more frequently than that? (possibly in other hooks as well as update-status hook)

**Comment by shayancanonical** [comment]
@theoctober19th usually `kubectl.apply` (or lightkube.apply) can take a long time (normally, about 5-10mins) to create a service of type `loadbalancer` on various cloud provides as the cloud provider needs to provision certain resources. as a result, the following is the intended behavior for this PR (the code as it currently exists will likely need to be modified to accomplish this):
1. upon `config-changed` we will make a request to K8s to create the service of the desired type if necessary. we will likely set the charm in `MaintenanceStatus` and return from the hook so other hooks can run
2. upon a future hook, the `reconcile` approach of this charm allows us to check if the K8s service created in (1) is reachable. if it is reachable, we will update the endpoints in the relation databag
3. while the K8s service is being created, we will avoid touching the databag at all. if the K8s service is being created for the first time, the `endpoints` in the databag will be empty. if not the first time, the `endpoints` will remain stale until the new K8s service is reachable (we can optimize here by making use of pebble notices for more frequent checks, or wait until the next hook - at most update-status-hook-interval)

**Comment by theoctober19th** [comment]
Thanks for the explanation, @shayancanonical.

We had discussed this in our team, and thanks to @welpaolo we had discussed some additional possibility of having a flag set to false in peer relation database whenever a service is being created / deleted, which would then trigger peer-relation-changed event, and in that event we check for the availablity of the service and either a) reset the flag, and update the endpoints or b) defer the peer-relation-changed event hook and then basically repeat this process when the deferred hook gets fired later.

This can also be combined with checking the status of service in other event hooks (including update-status hook). This will effectively check service availablity during either peer-relation-changed or other event hooks, whichever occurs early.

File: tests/unit/scenario_/database_relations/test_database_relations.py
Comment (shayancanonical) [reply tests/unit/scenario_/database_relations/test_database_relations.py:22]: it is because we're monkey patching `model_service_domain` above which will affect the output of `_get_hosts_ports`

File: src/relations/database_provides.py
Comment (shayancanonical) [reply src/relations/database_provides.py:186]: i think that this file has implementation details that significantly diverge from the vm implementation (how endpoints are determined). thus, we should not share the `data_provides.py` file between vm and k8s

furthermore, we should take a more intentional approach to shared code between the routers

File: config.yaml
Comment (sinclert-canonical) [reply config.yaml:8]: > default false may imply true is valid value. Change to something else? no none

👋🏻 I subscribe Paulo's comment here.

I understand we are "compromising" in order to have a similar interface across DP charms (i.e. [kafka-k8s](https://github.com/canonical/kafka-k8s-operator/blob/f4d9336d96fc7e4aa47a5afb112b5dbf8d5ade6e/config.yaml#L115-L118)), but it is confusing and we all should make an effort to reduce these occurrences in the future.

File: src/kubernetes_charm.py
Comment (sinclert-canonical) [inline src/kubernetes_charm.py:147]: Are we sure we could cache the result of this function? I am not worried about the _size_ of the cache, but the _validity_ of the result being cached.

Are we sure that, during the execution of an event hook, there can not be multiple units with the same name? If such scenario is imposible, shouldn't the argument be called `unit_id` instead? Not sure if that jargon is used somewhere else.

File: src/kubernetes_charm.py
Comment (sinclert-canonical) [inline src/kubernetes_charm.py:156]: Are we sure we could cache the result of this function? I am not worried about the _size_ of the cache, but the _validity_ of the result being cached.

Are we sure that, during the execution of an event hook, there can not be multiple units with the same name? If such scenario is imposible, shouldn't the argument be called `unit_id` instead? Not sure if that jargon is used somewhere else.

**Review by sinclert-canonical — APPROVED** [review]
For what I understand LGTM (I read [this spec](https://docs.google.com/document/d/1-R4hYgBbM77sarTegtKa_MymikKSnhYYB7Av3lWvjKw/edit?tab=t.0) to catch up with the context) 👍🏻 

I would suggest addressing all the unresolved comments before going forward with the merge. It seems there were many conversations before my arrival.

File: src/abstract_charm.py
Comment (shayancanonical) [reply src/abstract_charm.py]: this is indeed k8s only. have updated the comment to make this explicit

File: src/abstract_charm.py
Comment (shayancanonical) [reply src/abstract_charm.py]: opted to use an abstractmethod inhttps://github.com/canonical/mysql-router-k8s-operator/pull/328/commits/c6f686906759d0aa5072864b9c40fb1273d4aaa0 -> this will help abstract out different logic for vm and k8s

**Review by paulomach — APPROVED** [review]
Nice work. Let's get this merged

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py:156]: As paulo mentioned, unit_name is unique per unit and has a 1:1 mapping with the pod name

File: src/kubernetes_charm.py
Comment (shayancanonical) [reply src/kubernetes_charm.py]: addressed in [fb65030](https://github.com/canonical/mysql-router-k8s-operator/pull/328/commits/fb65030b489e492e36ccf3131c04fd55ec8ba72d)

File: src/kubernetes_charm.py
Comment (sinclert-canonical) [reply src/kubernetes_charm.py:147]: All clear. Unit names are preserve uniqueness.

**Pull request merged**
